File: solutions/python/07/discrete_log/main.py
def bsgs(a, b, p, m):
    a %= p
    b %= p
    step = int(m**0.5) + 1
    baby = {1: 0}
    cur = 1
    for j in range(1, step + 1):
        cur = cur * a % p
        if cur not in baby:
            baby[cur] = j
    factor = pow(a, step * (p - 2), p)
    gamma = b
    for i in range(step + 1):
        if gamma in baby:
            return i * step + baby[gamma]
        gamma = gamma * factor % p
    return -1


def main() -> None:
    p = 1_000_000_009
    a, b = 2, 8
    print(bsgs(a, b, p, p - 1))


if __name__ == "__main__":
    main()

File: solutions/go/07/discrete_log/main.go
package main

import (
	"fmt"
	"math"
)

func bsgs(a, b, p, m int) int {
	a %= p
	b %= p
	step := int(math.Sqrt(float64(m))) + 1
	baby := map[int]int{1: 0}
	cur := 1
	for j := 1; j <= step; j++ {
		cur = cur * a % p
		if _, ok := baby[cur]; !ok {
			baby[cur] = j
		}
	}
	factor := powMod(a, step*(p-2), p)
	gamma := b
	for i := 0; i <= step; i++ {
		if j, ok := baby[gamma]; ok {
			return i*step + j
		}
		gamma = gamma * factor % p
	}
	return -1
}

func powMod(a, e, p int) int {
	res := 1
	a %= p
	for e > 0 {
		if e&1 == 1 {
			res = res * a % p
		}
		a = a * a % p
		e >>= 1
	}
	return res
}

func main() {
	p := 1_000_000_009
	fmt.Println(bsgs(2, 8, p, p-1))
}
